mozart++/system/process.hpp

```diff
@@ -0,0 +1,504 @@
+                    close_fd(pstderr[PIPE_READ]);
+                }
+                dup2(pstderr[PIPE_WRITE], STDERR_FILENO);
+            }
+
+            close_fd(pstdin[PIPE_READ]);
+            close_fd(pstdout[PIPE_WRITE]);
+            close_fd(pstderr[PIPE_WRITE]);
+
+            // copy command-line arguments
+            size_t asize = startup._cmdline.size();
+            char *argv[asize + 1];
+
+            // argv is always terminated with a nullptr
+            argv[asize] = nullptr;
+
+            for (std::size_t i = 0; i < asize; ++i) {
+                argv[i] = strdup(startup._cmdline[i].c_str());
+            }
+
+            // change cwd
+            if (chdir(startup._cwd.c_str()) != 0) {
+                mpp::throw_ex<mpp::runtime_error>("unable to change current working directory");
+            }
+
+            // copy environment variables
+            size_t esize = startup._env.size();
+            char *envp[esize + 1];
+
+            // for safety, envp will be terminated with a nullptr
+            envp[esize] = nullptr;
+
+            std::stringstream buffer;
+            char **penv = envp;
+            for (const auto &e : startup._env) {
+                buffer.str("");
+                buffer.clear();
+                buffer << e.first << "=" << e.second;
+                *penv++ = strdup(buffer.str().c_str());
+            }
+
+            // run subprocess
+            ::execve(argv[0], argv, envp);
+
+            // throw if failed to execve()
+            for (std::size_t i = 0; i < asize; ++i) {
+                ::free(argv[i]);
+            }
+            for (std::size_t i = 0; i < esize; ++i) {
+                ::free(envp[i]);
+            }
+            mpp::throw_ex<mpp::runtime_error>("unable to exec commands in subprocess");
+
+        } else {
+            // in parent process
+            if (!startup._stdin.redirected()) {
+                close_fd(pstdin[PIPE_READ]);
+            }
+            if (!startup._stdout.redirected()) {
+                close_fd(pstdout[PIPE_WRITE]);
+            }
+
+            /*
+             * pay special attention to stderr,
+             * there are 2 cases:
+             *      1. redirect stderr to stdout
+             *      2. redirect stderr to a file
+             */
+            if (startup._redirect_error) {
+                // redirect stderr to stdout
+                // do nothing
+            } else {
+                // redirect stderr to a file
+                if (!startup._stderr.redirected()) {
+                    close_fd(pstderr[PIPE_WRITE]);
+                }
+            }
+
+            info._pid = pid;
+            info._stdin = pstdin[PIPE_WRITE];
+            info._stdout = pstdout[PIPE_READ];
+            info._stderr = pstderr[PIPE_READ];
+
+            // on *nix systems, fork() doesn't create threads to run process
+            info._tid = FD_INVALID;
+        }
+    }
+
+#endif
```

Comment:
In fact, considering the possibilities of throwing an other exception inside this function(such as constructor of std::string), I strongly recommend using RAII to protect pipes here.

Reply:
It's not easy to apply RAII to these pipe fds. 
First of all, sometimes fds[0] and fds[1] should be both closed (when errored), but sometimes one of fds[0] and fds[1] should remain open util process exits.

The best way to ensure no resource leak is applying try-catch in its caller and close pipes.

Reply:
ok. that is better.